docs/git-guide.md

```diff
@@ -0,0 +1,1052 @@
+
+First, sign in to [Travis CI][travis-ci] with your GitHub account and authorize Travis CI to access your GitHub account and repositories. Once you've done this, Travis CI will fetch your repository information and display it on your [profile page][travis-ci-profile]. From there you can enable integration with Zulip. ([See screencast](_static/zulip-travisci.gif).)
+
+## Contribute
+
+### Know what branch you're working on
+
+When using Git, it's important to know which branch you currently have checked out because most git commands implicitly operate on the current branch. You can determine the currently checked out branch several ways.
+
+One way is with [git status][gitbook-git-status]:
+
+```
+$ git status
+On branch issue-demo
+nothing to commit, working directory clean
+```
+
+Another is with [git branch][gitbook-git-branch] which will display all local branches, with a star next to the current branch:
+
+```
+$ git branch
+* issue-demo
+  master
+```
+
+To see even more information about your branches, including remote branches, use `git branch -vva`:
+
+```
+$ git branch -vva
+* issue-123                 517468b troubleshooting tip about provisioning
+  master                    f0eaee6 [origin/master] bug: Fix traceback in get_missed_message_token_from_address().
+  remotes/origin/HEAD       -> origin/master
+  remotes/origin/issue-1234 4aeccb7 Another test commit, with longer message.
+  remotes/origin/master     f0eaee6 bug: Fix traceback in get_missed_message_token_from_address().
+  remotes/upstream/master   dbeab6a Optimize checks of test database state by moving into Python.
+```
+
+You can also configure [Bash][gitbook-other-envs-bash] and [Zsh][gitbook-other-envs-zsh] to display the current branch in your prompt.
+
+### Keep your fork up to date
+
+You'll want to [keep your fork][github-help-sync-fork] up-to-date with changes from Zulip's main repositories.
+
+**Note about git pull**: You might be used to using `git pull` on other projects. With Zulip, because we don't use merge commits, you'll want to avoid it. Rather that using `git pull`, which by default is a shortcut for `git fetch && git merge FECH_HEAD` ([docs][gitbook-git-pull]), you should use `git fetch` and then `git rebase`.
+
+First, [fetch][gitbook-fetch] changes from Zulip's upstream repository you configured in the step above:
+
+```
+$ git fetch upstream
+```
+
+Next, checkout your `master` branch and [rebase][gitbook-git-rebase] it on top of `upstream/master`:
+
+```
+$ git checkout master
+Switched to branch 'master'
+
+$ git rebase upstream/master
+```
+
+This will rollback any changes you've made to master, update it from `upstream/master`, and then re-apply your changes. Rebasing keeps the commit history clean and readable.
+
+When you're ready, [push your changes][github-help-push] to your remote fork. Make sure you're in branch `master` and the run `git push`:
+
+```
+$ git checkout master
+$ git push origin master
```

Comment:
Is there any significant difference between `+` and `--force`? 

I got into use of `--force` later on, but I will include some of that in this section.